kernel/src/process/signal/poll.rs

```diff
@@ -131,75 +126,138 @@ impl Pollee {
     }
 }
 
-/// A poller gets notified when its associated pollees have interesting events.
-pub struct Poller {
-    // Use event counter to wait or wake up a poller
-    event_counter: Arc<EventCounter>,
-    // All pollees that are interesting to this poller
-    pollees: Vec<Weak<PolleeInner>>,
-    // A waiter used to pause the current thread.
-    waiter: Waiter,
+/// A poller that is notified when its associated pollees have interesting events.
+///
+/// The observer type, `T`, should be a type that implements the [`Observer`] trait.
+pub struct Poller<T> {
+    // The event observer.
+    observer: Arc<T>,
+    // The inner with observer type erased.
+    inner: AnyPoller,
 }
 
-impl Default for Poller {
-    fn default() -> Self {
-        Self::new()
+impl<T: Observer<IoEvents> + 'static> Poller<T> {
+    /// Constructs a new poller with the specified observer.
+    pub fn with_observer(observer: T) -> Self {
```

Comment:
have some skepticism about whether users would want to construct a `Poller` instead of a `WaitablePoller`. After searching this PR, I found that `Poller::with_observer` is called only once (and not within `WaitablePoller`) in the `set_owner` method. Additionally, if we use `WaitablePoller`, it seems that there would be no drawbacks. What if we simply expose `WaitablePoller` to users and keep `Poller` as an internal API? In that case, it might make more sense to rename `WaitablePoller` to just `Poller` for the sake of user simplicity.

Reply:
> I found that `Poller::with_observer` is called only once

`Pollers` are designed to be used in the epoll file and in the `SIGIO` delivery. However, in this PR, the epoll file uses `AnyPoller` directly. In principle, it should use `Poller` instead (we may want to discuss this in https://github.com/asterinas/asterinas/pull/1500).

> Additionally, if we use `WaitablePoller`, it seems that there would be no drawbacks.

I do not see how this is possible. A `Poller` can be with any observer, and the observer may not support waiting for events. This is the case for the epoll file and the `SIGIO` package.

> In that case, it might make more sense to rename `WaitablePoller` to just `Poller` for the sake of user simplicity.

Since `WaitablePoller` is used more often, I agree that it deserves a short name. A possible way is to rename:
 - `WaitablePoller` -> `Poller`
 - `Poller` -> `RawPoller`

Reply:
> I do not see how this is possible. A Poller can be with any observer, and the observer may not support waiting for events. This is the case for the epoll file and the SIGIO delivery.

I misunderstand the meaning of `Poller`. Then your use should be correct.

> Since WaitablePoller is used more often, I agree that it deserves a short name. A possible > > way is to rename:

> - WaitablePoller -> Poller
> - Poller -> RawPoller

Yes, the renaming will be better.

Reply:
Renamed and also updated the PR description.